Cython/Compiler/Nodes.py

```diff
@@ -6999,36 +7020,42 @@ def generate_execution_code(self, code):
                     lhs_pos=value.pos,
                     rhs=value,
                     code=code,
-                    have_gil=self.in_nogil_context)
-                value.generate_post_assignment_code(code)
+                    have_gil=code.funcstate.gil_owned)
             else:
                 value.make_owned_reference(code)
                 code.putln("%s = %s;" % (
                     Naming.retval_cname,
                     value.result_as(self.return_type)))
-                value.generate_post_assignment_code(code)
-                if code.globalstate.directives['profile'] or code.globalstate.directives['linetrace']:
-                    code.put_trace_return(
-                        Naming.retval_cname,
-                        self.pos,
-                        return_type=self.return_type,
-                        nogil=not code.funcstate.gil_owned,
-                    )
-            value.free_temps(code)
         else:
             if self.return_type.is_pyobject:
                 code.put_init_to_py_none(Naming.retval_cname, self.return_type)
             elif self.return_type.is_returncode:
                 self.put_return(code, self.return_type.default_value)
 
+        code.putln("}")  # end of omp critical section
+        if self.return_type.needs_refcounting:
+            code.put_xdecref(
+                Naming.quick_temp_cname, self.return_type, have_gil=code.funcstate.gil_owned)
+            code.putln("}")  # end of quick_temp scope
+
+        if value:
+            value.generate_post_assignment_code(code)
+            if (not self.return_type.is_memoryviewslice and
+                    code.globalstate.directives['profile'] or code.globalstate.directives['linetrace']):
+                code.put_trace_return(
+                    Naming.retval_cname,
+                    self.pos,
+                    return_type=self.return_type,
+                    nogil=not code.funcstate.gil_owned,
+                )
```

Comment:
I need to have a second look at this, but tracing feels like it might need some thread safety too. Although maybe that's a whole, much broader question.

Reply:
For now I've decided to just disable this in parallel blocks. I think it's such a corner-case that it's barely worth the complication.